pennylane/devices/experimental/device_interface.py

```diff
@@ -0,0 +1,234 @@
+
+        return qscript, blank_postprocessing_fn
+
+    @abc.abstractmethod
+    def execute(
+        self, qscripts: Union[QuantumScript, QScriptBatch], execution_config: ExecutionConfig = None
+    ):
+        """Execute a Quantum Script and turn it into a result.
+
+        Args:
+            qscripts (Union[QuantumScript, Sequence[QuantumScript]]): the QuantumScript to be executed
+            execution_config (ExecutionConfig): a datastructure with all additional information required for execution
+
+        Returns:
+            A numeric result of the computation.
+
+        **Return shape:**
+
+        The result for each :class:`~.QuantumScript` must match the shape specified by :class:`~.QuantumScript.shape`.
+
+        For example:
+
+        >>> from pennylane.devices import experimental as devices
+        >>> qml.enable_return()
+        >>> dev = devices.PythonDevice()
+        >>> dev = devices.backward_patch_interface(dev)
+        >>> qs = qml.tape.QuantumScript([], [qml.expval(qml.PauliZ(0)), qml.probs(wires=(0,1))])
+        >>> qs.shape(dev)
+        ((), (4,))
+        >>> dev.execute(qs)
+        (1.0, array([1., 0., 0., 0.]))
+
+        """
+        raise NotImplementedError
+
+    def execute_and_gradients(
+        self,
+        qscripts: Union[QuantumScript, QScriptBatch],
+        execution_config: ExecutionConfig = None,
+    ):
+        """Compute the results and gradients of QuantumScripts at the same time.
+
+        Args:
+            qscripts (Union[QuantumScript, Sequence[QuantumScript]]): the QuantumScript to be executed
+            execution_config (ExecutionConfig): a datastructure with all additional information required for execution
+
+        Returns:
+            tensor-like, tensor-like: A numeric result of the computation.
+
+        This method can be used when the result and execution need to be computed at the same time, such as
+        during a forward mode calculation of gradients. For certain gradient methods, such as adjoint
+        diff gradients, calculating the result and gradient at the same can save computational work.
+
+        """
+        return self.execute(qscripts, execution_config), self.gradient(qscripts, execution_config)
+
+    def gradient(
+        self,
+        qscript: Union[QuantumScript, QScriptBatch],
+        execution_config: ExecutionConfig = None,
+    ):
+        """Calculate the gradient of either a single or a batch of Quantum Scripts.
+
+        Args:
+            qscripts (Union[QuantumScript, Sequence[QuantumScript]]): the QuantumScript to be executed
+            execution_config (ExecutionConfig): a datastructure with all additional information required for execution
+
+        Returns:
+            tensor-like: A numeric result of the computation.
+        """
+        raise NotImplementedError
+
+    def vjp(
+        self,
+        qscript: Union[QuantumScript, QScriptBatch],
+        dy,
+        execution_config: ExecutionConfig = None,
+    ):
```

Comment:
@mlxd this wasn't evident when we wrote the ADR originally, but since then new research has come out, and it might be good to also have an optional `jvp` method. For example, there may be a possibility to extract JVPs directly from quantum hardware, in a manner that is more efficient than computing the full gradient.

Reply:
Sure, I think we can work that in as an optional method.